File: MediTrackWeb/src/services/authService.js
import apiService from './api';

class AuthService {
  // Sign up new user
  async signUp(userData) {
    try {
      const response = await apiService.post('/api/auth/signup', userData, false);
      return response;
    } catch (error) {
      throw new Error(error.message || 'Sign up failed');
    }
  }

  // Sign in user
  async signIn(email, password) {
    try {
      const response = await apiService.post('/api/auth/signin', {
        email,
        password
      }, false);

      if (response.session?.access_token) {
        apiService.setToken(response.session.access_token);
        // Store user data
        localStorage.setItem('user_data', JSON.stringify({
          user: response.user,
          profile: response.profile
        }));
      }

      return response;
    } catch (error) {
      throw new Error(error.message || 'Sign in failed');
    }
  }

  // Sign out user
  async signOut() {
    try {
      await apiService.post('/api/auth/signout', {});
      this.clearUserData();
      return { success: true };
    } catch (error) {
      // Even if API call fails, clear local data
      this.clearUserData();
      throw new Error(error.message || 'Sign out failed');
    }
  }

  // Verify token
  async verifyToken() {
    try {
      const response = await apiService.get('/api/auth/verify');
      return response;
    } catch (error) {
      this.clearUserData();
      throw new Error('Token verification failed');
    }
  }

  // Reset password
  async resetPassword(email) {
    try {
      const response = await apiService.post('/api/auth/reset-password', { email }, false);
      return response;
    } catch (error) {
      throw new Error(error.message || 'Password reset failed');
    }
  }

  // Update password
  async updatePassword(newPassword, refreshToken) {
    try {
      const response = await apiService.post('/api/auth/update-password', {
        new_password: newPassword,
        refresh_token: refreshToken
      });
      return response;
    } catch (error) {
      throw new Error(error.message || 'Password update failed');
    }
  }

  // Get current user data from localStorage
  getCurrentUser() {
    try {
      const userData = localStorage.getItem('user_data');
      return userData ? JSON.parse(userData) : null;
    } catch (error) {
      console.error('Error getting user data:', error);
      return null;
    }
  }

  // Check if user is authenticated
  isAuthenticated() {
    const token = localStorage.getItem('auth_token');
    const userData = this.getCurrentUser();
    return !!(token && userData);
  }

  // Clear user data from localStorage
  clearUserData() {
    localStorage.removeItem('auth_token');
    localStorage.removeItem('user_data');
    apiService.setToken(null);
  }

  // Get user role
  getUserRole() {
    const userData = this.getCurrentUser();
    return userData?.profile?.role || 'patient';
  }

  // Auto-login check (verify token on app load)
  async autoLogin() {
    const token = localStorage.getItem('auth_token');
    if (!token) {
      return false;
    }

    try {
      await this.verifyToken();
      return true;
    } catch (error) {
      this.clearUserData();
      return false;
    }
  }
}

export default new AuthService();

File: MediTrack/BackEnd/services/auth_service.py
from supabase import create_client, Client
from jose import JWTError, jwt
from typing import Optional, Dict, Any
import httpx
from config.settings import Settings

class AuthService:
    def __init__(self):
        self.settings = Settings()
        self.client: Client = create_client(
            self.settings.supabase_url,
            self.settings.supabase_key
        )
    
    async def verify_token(self, token: str) -> Dict[str, Any]:
        """Verify JWT token and return user data"""
        try:
            # Verify with Supabase
            response = self.client.auth.get_user(token)
            if response.user:
                return {
                    "id": response.user.id,
                    "email": response.user.email,
                    "user_metadata": response.user.user_metadata
                }
            else:
                raise Exception("Invalid token")
        except Exception as e:
            raise Exception(f"Token verification failed: {str(e)}")
    
    async def sign_up(self, email: str, password: str, user_data: Dict[str, Any] = None) -> Dict[str, Any]:
        """Sign up new user"""
        try:
            response = self.client.auth.sign_up({
                "email": email,
                "password": password,
                "options": {
                    "data": user_data or {}
                }
            })
            
            if response.user:
                return {
                    "user": {
                        "id": response.user.id,
                        "email": response.user.email,
                        "email_confirmed_at": response.user.email_confirmed_at
                    },
                    "session": {
                        "access_token": response.session.access_token if response.session else None,
                        "refresh_token": response.session.refresh_token if response.session else None
                    }
                }
            else:
                raise Exception("Sign up failed")
                
        except Exception as e:
            raise Exception(f"Sign up error: {str(e)}")
    
    async def sign_in(self, email: str, password: str) -> Dict[str, Any]:
        """Sign in user"""
        try:
            response = self.client.auth.sign_in_with_password({
                "email": email,
                "password": password
            })
            
            if response.user and response.session:
                return {
                    "user": {
                        "id": response.user.id,
                        "email": response.user.email,
                        "last_sign_in_at": response.user.last_sign_in_at
                    },
                    "session": {
                        "access_token": response.session.access_token,
                        "refresh_token": response.session.refresh_token,
                        "expires_at": response.session.expires_at
                    }
                }
            else:
                raise Exception("Invalid credentials")
                
        except Exception as e:
            raise Exception(f"Sign in error: {str(e)}")
    
    async def refresh_token(self, refresh_token: str) -> Dict[str, Any]:
        """Refresh access token"""
        try:
            response = self.client.auth.refresh_session(refresh_token)
            
            if response.session:
                return {
                    "access_token": response.session.access_token,
                    "refresh_token": response.session.refresh_token,
                    "expires_at": response.session.expires_at
                }
            else:
                raise Exception("Token refresh failed")
                
        except Exception as e:
            raise Exception(f"Token refresh error: {str(e)}")
    
    async def sign_out(self, token: str) -> bool:
        """Sign out user"""
        try:
            self.client.auth.sign_out()
            return True
        except Exception as e:
            return False
    
    async def reset_password(self, email: str) -> bool:
        """Send password reset email"""
        try:
            response = self.client.auth.reset_password_email(email)
            return True
        except Exception as e:
            raise Exception(f"Password reset error: {str(e)}")
    
    async def update_password(self, refresh_token: str, new_password: str) -> bool:
        """Update user password"""
        try:
            # Set the session
            #self.client.auth.set_session(token, None)
            session = self.client.auth.refresh_session(refresh_token)
            response = self.client.auth.update_user({
                "password": new_password
            })
            return response.user is not None
        except Exception as e:
            raise Exception(f"Password update error: {str(e)}")
    
    def extract_user_id_from_token(self, token: str) -> str:
        """Extract user ID from JWT token without verification (for internal use)"""
        try:
            payload = jwt.get_unverified_claims(token)
            return payload.get("sub")
        except JWTError:
            raise Exception("Invalid token format")